core/src/main/scala/cats/data/RWST.scala

```diff
@@ -0,0 +1,550 @@
+package cats
+package data
+
+import cats.functor.{ Contravariant, Bifunctor, Profunctor }
+import cats.syntax.either._
+
+/**
+  * Represents a stateful computation in a context `F[_]`, over state `S`, with an initial environment `E`, an accumulated log `L` and a result `A`.
+  *
+  * In other words, it is a pre-baked stack of `[[ReaderT]][F, E, A]`, `[[WriterT]][F, L, A]`
+  * and `[[StateT]][F, S, A]`.
+  */
+final class RWST[F[_], E, S, L, A](val runF: F[(E, S) => F[(L, S, A)]]) extends Serializable {
+
+  /**
+    * Modify the initial environment using `f`.
+    */
+  def contramap[E0](f: E0 => E)(implicit F: Functor[F]): RWST[F, E0, S, L, A] =
+    RWST.applyF {
+      F.map(runF) { rwsa =>
+        (e0: E0, s: S) => rwsa(f(e0), s)
+      }
+    }
+
+  /**
+    * Alias for [[contramap]].
+    */
+  def local[EE](f: EE => E)(implicit F: Functor[F]): RWST[F, EE, S, L, A] =
+    contramap(f)
+
+  /**
+    * Modify the result of the computation using `f`.
+    */
+  def map[B](f: A => B)(implicit F: Functor[F]): RWST[F, E, S, L, B] =
+    transform { (l, s, a) => (l, s, f(a)) }
+
+  /**
+    * Modify the written log value using `f`.
+    */
+  def mapWritten[LL](f: L => LL)(implicit F: Functor[F]): RWST[F, E, S, LL, A] =
+    transform { (l, s, a) => (f(l), s, a) }
+
+  /**
+    * Combine this computation with `rwsb` using `fn`. The state will be be threaded
+    * through the computations and the log values will be combined.
+    */
+  def map2[B, Z](rwsb: RWST[F, E, S, L, B])(fn: (A, B) => Z)(implicit F: FlatMap[F], L: Semigroup[L]): RWST[F, E, S, L, Z] =
```

Comment:
Is there an advantage to having this as a separate implementation, rather than just straight delegating to `flatMap`?